File: mnheme_ports/java/mnheme/IndexEngine.java
// mnheme/java/mnheme/IndexEngine.java
package mnheme;

import java.util.*;
import java.util.regex.*;
import java.util.stream.*;

/**
 * In-RAM index engine. Rebuilt from log on cold start.
 *
 * concept_index : Map<String, List<Long>>
 * feeling_index : Map<String, List<Long>>
 * cf_index      : Map<"concept\0feeling", List<Long>>
 * word_index    : Map<String, List<Long>>  -- inverted full-text
 */
public class IndexEngine {

    private static final Pattern TOKEN_RE =
        Pattern.compile("[a-zA-Zàáâäãåæçèéêëìíîïòóôöõùúûüýÿñ]{3,}",
                        Pattern.UNICODE_CHARACTER_CLASS);

    private final Map<String, List<Long>> concept    = new HashMap<>();
    private final Map<String, List<Long>> feeling    = new HashMap<>();
    private final Map<String, List<Long>> tag        = new HashMap<>();
    private final Map<String, List<Long>> cf         = new HashMap<>();
    private final List<String[]>          timelineTs = new ArrayList<>();
    private final List<Long>              all        = new ArrayList<>();
    private final Map<String, List<Long>> word       = new HashMap<>();

    public void indexRecord(long offset, Map<String, Object> r) {
        String c  = strVal(r, "concept");
        String f  = strVal(r, "feeling");
        String ts = strVal(r, "timestamp");

        concept.computeIfAbsent(c, k -> new ArrayList<>()).add(offset);
        feeling.computeIfAbsent(f, k -> new ArrayList<>()).add(offset);
        cf.computeIfAbsent(cfKey(c, f), k -> new ArrayList<>()).add(offset);
        timelineTs.add(new String[]{ts, String.valueOf(offset)});
        all.add(offset);

        Object tagsObj = r.get("tags");
        if (tagsObj instanceof List<?> tags) {
            for (Object t : tags) {
                if (t instanceof String s && !s.isBlank())
                    tag.computeIfAbsent(s, k -> new ArrayList<>()).add(offset);
            }
        }

        String combined = strVal(r, "content") + " " + c + " " + strVal(r, "note");
        for (String tok : tokenize(combined))
            word.computeIfAbsent(tok, k -> new ArrayList<>()).add(offset);
    }

    public int rebuild(List<StorageEngine.ScanResult> records) {
        concept.clear(); feeling.clear(); tag.clear();
        cf.clear(); timelineTs.clear(); all.clear(); word.clear();
        for (var sr : records) indexRecord(sr.offset(), sr.record());
        timelineTs.sort(Comparator.comparing(a -> a[0]));
        return records.size();
    }

    public List<Long> offsetsByConcept(String c, String f, boolean oldestFirst) {
        List<Long> offs = new ArrayList<>(
            f != null && !f.isEmpty()
                ? cf.getOrDefault(cfKey(c, f), List.of())
                : concept.getOrDefault(c, List.of()));
        if (!oldestFirst) Collections.reverse(offs);
        return offs;
    }

    public List<Long> offsetsByFeeling(String f, boolean oldestFirst) {
        List<Long> offs = new ArrayList<>(feeling.getOrDefault(f, List.of()));
        if (!oldestFirst) Collections.reverse(offs);
        return offs;
    }

    public List<Long> allOffsets(boolean oldestFirst) {
        List<Long> offs = new ArrayList<>(all);
        if (!oldestFirst) Collections.reverse(offs);
        return offs;
    }

    public List<Long> timelineOffsets(String c) {
        Set<Long> cset = new HashSet<>(concept.getOrDefault(c, List.of()));
        return timelineTs.stream()
            .map(a -> Long.parseLong(a[1]))
            .filter(cset::contains)
            .collect(Collectors.toList());
    }

    public List<String> allConcepts() {
        return concept.keySet().stream().sorted().collect(Collectors.toList());
    }

    public int count(String c, String f) {
        if (c != null && f != null) return cf.getOrDefault(cfKey(c, f), List.of()).size();
        if (c != null)               return concept.getOrDefault(c, List.of()).size();
        if (f != null)               return feeling.getOrDefault(f, List.of()).size();
        return all.size();
    }

    public Map<String, Integer> feelingDistribution() {
        Map<String, Integer> dist = new LinkedHashMap<>();
        feeling.forEach((k, v) -> dist.put(k, v.size()));
        return dist;
    }

    public Map<String, Map<String, Integer>> conceptFeelingMatrix() {
        Map<String, Map<String, Integer>> matrix = new LinkedHashMap<>();
        cf.forEach((key, offsets) -> {
            int idx = key.indexOf('\0');
            String cKey = key.substring(0, idx);
            String fKey = key.substring(idx + 1);
            matrix.computeIfAbsent(cKey, k -> new LinkedHashMap<>()).put(fKey, offsets.size());
        });
        return matrix;
    }

    private static String cfKey(String c, String f) { return c + '\0' + f; }

    private static Set<String> tokenize(String text) {
        Set<String> tokens = new HashSet<>();
        Matcher m = TOKEN_RE.matcher(text.toLowerCase(Locale.ROOT));
        while (m.find()) tokens.add(m.group());
        return tokens;
    }

    private static String strVal(Map<String, Object> m, String key) {
        Object v = m.get(key);
        return v instanceof String s ? s : "";
    }
}

File: mnheme_ports/kotlin/IndexEngine.kt
// mnheme/kotlin/mnheme/IndexEngine.kt
package mnheme

import org.json.JSONObject

/**
 * In-RAM index engine — rebuilt from log on cold start.
 * Uses only stdlib + org.json (ships with Android).
 */
class IndexEngine {

    private val concept  = mutableMapOf<String, MutableList<Long>>()
    private val feeling  = mutableMapOf<String, MutableList<Long>>()
    private val tag      = mutableMapOf<String, MutableList<Long>>()
    private val cf       = mutableMapOf<Pair<String,String>, MutableList<Long>>()
    private val timeline = mutableListOf<Pair<String, Long>>() // (timestamp, offset) sorted
    private val all      = mutableListOf<Long>()
    private val word     = mutableMapOf<String, MutableList<Long>>() // inverted full-text

    fun indexRecord(offset: Long, r: JSONObject) {
        val c  = r.optString("concept")
        val f  = r.optString("feeling")
        val ts = r.optString("timestamp")

        concept .getOrPut(c) { mutableListOf() }.add(offset)
        feeling .getOrPut(f) { mutableListOf() }.add(offset)
        cf      .getOrPut(c to f) { mutableListOf() }.add(offset)
        timeline.add(ts to offset)
        all     .add(offset)

        val tags = r.optJSONArray("tags")
        if (tags != null) {
            for (i in 0 until tags.length()) {
                val t = tags.optString(i)
                if (t.isNotBlank()) tag.getOrPut(t) { mutableListOf() }.add(offset)
            }
        }

        // Inverted full-text
        val combined = "${r.optString("content")} $c ${r.optString("note")}"
        tokenize(combined).forEach { tok ->
            word.getOrPut(tok) { mutableListOf() }.add(offset)
        }
    }

    fun rebuild(records: List<StorageEngine.ScanResult>): Int {
        concept.clear(); feeling.clear(); tag.clear()
        cf.clear(); timeline.clear(); all.clear(); word.clear()
        records.forEach { indexRecord(it.offset, it.record) }
        timeline.sortBy { it.first }
        return records.size
    }

    // ── Queries ──────────────────────────────────────────────────

    fun offsetsByConcept(c: String, f: String?, oldestFirst: Boolean): List<Long> {
        val offs = if (f != null && f.isNotEmpty())
            cf[c to f]?.toMutableList() ?: mutableListOf()
        else
            concept[c]?.toMutableList() ?: mutableListOf()
        return if (oldestFirst) offs else offs.asReversed()
    }

    fun offsetsByFeeling(f: String, oldestFirst: Boolean): List<Long> {
        val offs = feeling[f]?.toMutableList() ?: mutableListOf()
        return if (oldestFirst) offs else offs.asReversed()
    }

    fun allOffsets(oldestFirst: Boolean): List<Long> {
        val offs = all.toMutableList()
        return if (oldestFirst) offs else offs.asReversed()
    }

    fun timelineOffsets(c: String): List<Long> {
        val cset = concept[c]?.toHashSet() ?: emptySet()
        return timeline.filter { it.second in cset }.map { it.second }
    }

    // ── Stats ─────────────────────────────────────────────────────

    fun allConcepts(): List<String> = concept.keys.sorted()

    fun count(c: String?, f: String?): Int = when {
        c != null && f != null -> cf[c to f]?.size ?: 0
        c != null              -> concept[c]?.size ?: 0
        f != null              -> feeling[f]?.size ?: 0
        else                   -> all.size
    }

    fun feelingDistribution(): Map<String, Int> =
        feeling.mapValues { it.value.size }

    fun conceptFeelingMatrix(): Map<String, Map<String, Int>> {
        val matrix = mutableMapOf<String, MutableMap<String, Int>>()
        cf.forEach { (key, offs) ->
            matrix.getOrPut(key.first) { mutableMapOf() }[key.second] = offs.size
        }
        return matrix
    }

    companion object {
        private val TOKEN_RE = Regex("[a-zA-Zàáâäãåæçèéêëìíîïòóôöõùúûüýÿñ]{3,}")

        fun tokenize(text: String): Set<String> =
            TOKEN_RE.findAll(text.lowercase()).map { it.value }.toHashSet()
    }
}
